summingbird-storm/src/main/scala/com/twitter/summingbird/storm/collector/AggregatorOutputCollector.scala

```diff
@@ -0,0 +1,103 @@
+package com.twitter.summingbird.storm.collector
+
+import com.twitter.algebird.Semigroup
+import com.twitter.summingbird.online.executor.KeyValueShards
+import com.twitter.summingbird.online.option.SummerBuilder
+import backtype.storm.spout.SpoutOutputCollector
+import com.twitter.algebird.util.summer.AsyncSummer
+import com.twitter.util.{ Await, Future }
+import java.util.{ List => JList }
+import scala.collection.mutable.{ Map => MMap }
+import scala.collection.mutable.{ MutableList => MList }
+import scala.collection.{ Map => CMap }
+import scala.collection.JavaConverters._
+import java.util.{ List => JList }
+
+class AggregatorOutputCollector[K, V: Semigroup](in: SpoutOutputCollector, func: JList[AnyRef] => JList[AnyRef], summerBuilder: SummerBuilder, summerShards: KeyValueShards) extends TransformingOutputCollector(in, func) {
+
+  var spoutCaches = MMap[String, AsyncSummer[(K, V), Map[K, V]]]()
+  var lastDump = System.currentTimeMillis()
+  var streamMessageIdTracker = MMap[String, MMap[Int, MList[Object]]]()
+
+  def timerFlush() = {
+    spoutCaches.keys.foreach { stream =>
+      val tupsOut = spoutCaches(stream).tick.map { convertToSummerInputFormat(_) }
+      val tups = Await.result(tupsOut)
+
+      tups.foreach {
+        case (k, v) => {
+          val messageIdsTracker = streamMessageIdTracker(stream)
+          val messageIds = messageIdsTracker.remove(k)
+          (messageIds.isEmpty, stream.isEmpty) match {
+            case (true, true) => in.emit(List(k, v).asJava.asInstanceOf[JList[AnyRef]])
+            case (true, false) => in.emit(stream, List(k, v).asJava.asInstanceOf[JList[AnyRef]])
+            case (false, true) => in.emit(List(k, v).asJava.asInstanceOf[JList[AnyRef]], messageIds)
+            case (false, false) => in.emit(stream, List(k, v).asJava.asInstanceOf[JList[AnyRef]], messageIds)
+          }
+        }
+      }
+    }
+  }
+
+  private def convertToSummerInputFormat(data: CMap[K, V]): CMap[Int, CMap[K, V]] = {
+    data.groupBy { case (k, _) => summerShards.summerIdFor(k) }
+  }
+
+  private def emitData[K, V](data: Future[Traversable[(Int, CMap[K, V])]], callerfunc: (String, JList[AnyRef], scala.Any) => JList[Integer], s: String, o: scala.Any): List[Int] = {
+    var returns = MList[Int]()
+    val data_trav = Await.result(data)
+    data_trav
+      .foreach {
+        case (k, v) => {
+          val messageIdsTracker = streamMessageIdTracker(s)
+          val messageIds = messageIdsTracker.remove(k)
+          val taskIds = callerfunc(s, List(k, v).asJava.asInstanceOf[JList[AnyRef]], messageIds)
+          if (taskIds != null) returns ++= taskIds.asInstanceOf[JList[Int]].asScala.toList
+        }
+      }
+    returns.toList
+  }
+
+  private def add(tuple: (K, V), o: scala.Any, s: String) = {
+    if (o != Nil) trackMessageId(tuple, o, s)
```

Comment:
Nil as a placeholder here is really messy (there's no indication that it's a better placeholder than, say, `None` or even `null`).  The two approaches I think worth considering are:

1. Make `o: Option[Any]`
1. Make an `emitHelper` variant that doesn't take `o`, and just calls `addToCache` directly